libinfinity/common/inf-file-util.c

```diff
@@ -517,4 +517,49 @@ inf_file_util_delete(const gchar* path,
     return inf_file_util_delete_file(path, error);
 }
 
+/**
+ * inf_file_write_data_with_perms:
+ * @filename: Filename of the file to be written to.
+ * @data: Data that should be written to file.
+ * @length: Length of data in bytes.
+ * @flags: File access mode flags of the file to be open()'ed.
+ * @mode: File permission bits of the file to be open()'ed.
+ * @error: Location to store error information, if any, or %NULL.
+ *
+ * Writes @data to the file pointed to by @filename. Files may be created with
+ * specified permissions (@mode). If the function fails, %FALSE is returned
+ * and @error is set.
+ *
+ * Returns: %TRUE on success or %FALSE on error.
+ */
+gboolean
+inf_file_write_data_with_perms(const gchar* filename,
+                               const void* data,
+                               size_t length,
+                               int flags,
+                               int mode,
+                               GError** error)
+{
+  int fd = g_open(filename, flags, mode);
+  if (fd == -1)
+  {
+    inf_file_util_set_error_from_errno(error, errno);
+    return FALSE;
+  }
+  ssize_t remaining = length;
+  while (remaining > 0)
+  {
+    ssize_t written = write(fd, data, remaining);
+    if (remaining == -1)
+    {
+      g_close(fd, error);
+      inf_file_util_set_error_from_errno(error, errno);
+      return FALSE;
+    }
+    remaining -= written;
+  }
+  g_close(fd, error);
```

Comment:
Similar issue here: if `g_close` fails, this would set `error`, but the function still returns `TRUE`, which is different behavior than documented. I'd suggest to return `FALSE` if `g_close` fails (and maybe `g_unlink` the file), assuming that the file was not fully written to disk.